### PR TITLE
[FLINK-8411] Don't allow null in ListState.add()/addAll()

### PR DESCRIPTION
**Pull request by aljoscha**
R: @StefanRRichter, @bowenli86 

It turns out that this is a bit trickier than assumed earlier: `ListState.addAll()` was not considered and also had inconsistent behaviour between state backends before.

### CODE REVIEW

**Comment by bowenli86** [comment]
LGTM

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java [hunk 1]
```diff
@@ -696,6 +697,7 @@ public void clear() {
 
 		@Override
 		public void add(T value) throws Exception {
+			Objects.requireNonNull(value, "You cannot add null to a ListState.");
```
Comment (zentol) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: Why not our Preconditions version?

Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: Will change

**Comment by aljoscha** [comment]
@zentol pushed a fixup. What about the rest of the PR?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingListState.java [hunk 2]
```diff
@@ -48,6 +49,7 @@
 
 	@Override
 	public void add(T value) throws Exception {
+		Preconditions.checkNotNull(value, "You cannot add null to a ListState.");
```
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingListState.java]: You should also change the `addAll()`  to check for `nulls` internally, as done in the `HeapListState`.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java [hunk 3]
```diff
@@ -696,6 +696,7 @@ public void clear() {
 
 		@Override
```
Comment (kl0u) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: Change also the `addAll()`.

Comment (kl0u) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: And also,  there is the `update()` method in the `ListState`  interface that should be changed in all implementations.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingListState.java (hunk 2 above)
Comment (aljoscha) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingListState.java]: Or not change this one, and defer null detection completely to the wrapping state. WDYT?

Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingListState.java]: Yes, that would be cleaner.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java (hunk 3 above)
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: fixing

**Comment by aljoscha** [comment]
@kl0u Thanks for the review, and very good catch about `update()`!

**Pull request closed without merging**